snapcraft/meta/appstream.py

```diff
@@ -102,7 +104,15 @@ def extract(relpath: str, *, workdir: str) -> Optional[ExtractedMetadata]:
     version = _get_latest_release_from_nodes(dom.findall("releases/release"))
     project_license = _get_value_from_xml_element(dom, "project_license")
     update_contact = _get_value_from_xml_element(dom, "update_contact")
-    contact = [update_contact] if update_contact else None
+    contact = None
+    if update_contact:
+        if validators.url(update_contact) or validators.email(update_contact):
+            contact = [update_contact]
+        else:
+            emit.progress(
+                f"Ignoring invalid url {update_contact!r} in update_contact from appstream metadata.",
```

Comment:
Can we read to `contact` directly using `_get_urls_from_xml_element` instead of assigning to `update_contact`, revalidating it, and then assigning to `contact`?

Reply:
Agreed, not critical but this could be simplified.

Reply:
Perhaps this?

```python
contact = _get_value_from_xml_element(dom, "update_contact")
if not validators.url(update_contact) and not validators.email(update_contact):
     contact = None
     emit.progress(
                f"Ignoring invalid url {update_contact!r} in update_contact from appstream metadata.",
                permanent=True,
    )
```

Reply:
`contact` is a `list[str]` to match the `project` model. So, can we really do so?

Reply:
Ohh, right.  And I misunderstood what @cmatsuoka was asking about.

@cmatsuoka, I don't think we can call `_get_urls_from_xml_element` for `update_contact` because that is a single string whereas the `url` fields in an appstream are lists of strings.

Is that correct @soumyaDghosh?

Reply:
Okay, I think this is possible,

```python
    contact = _get_value_from_xml_element(dom, "update_contact")
    if contact:
        if not validators.url(contact) and not validators.email(contact):
            contact = None
            emit.progress(
                f"Ignoring invalid url {update_contact!r} in update_contact from appstream metadata.",
                permanent=True,
            )
```

And then, while returning, we'll just do this

```python
return ExtractedMetadata(
        common_id=common_id,
        title=title,
        summary=summary,
        description=description,
        version=version,
        icon=icon,
        license=project_license,
        contact=[contact],
        issues=issues,
        donation=donation,
        website=website,
        source_code=source_code,
        desktop_file_paths=desktop_file_paths,
    )
 ```

Reply:
> Ohh, right.  And I misunderstood what @cmatsuoka was asking about.
> 
> @cmatsuoka, I don't think we can call `_get_urls_from_xml_element` for `update_contact` because that is a single string whereas the `url` fields in an appstream are lists of strings.
> 
> Is that correct @soumyaDghosh?

Yeah, that's right.

Reply:
I see.  I like your original implementation the best!  We'll see what @cmatsuoka thinks

Reply:
Ah I see the list issue. My concern was about having the same validation twice in the code, but given the list vs scalar issue we could leave it as is.